app/models/configuration_script_payload.rb

```diff
@@ -3,11 +3,25 @@ class ConfigurationScriptPayload < ConfigurationScriptBase
 
   belongs_to :configuration_script_source
 
+  validate :validate_credentials_payload, :if => :credentials_changed?
+
   def self.base_model
     ConfigurationScriptPayload
   end
 
   def run(*)
     raise NotImplementedError, _("run must be implemented in a subclass")
   end
+
+  def validate_credentials_payload
+    return if credentials.blank?
+
+    error   = "credentials must be a hash" unless credentials.kind_of?(Hash)
+    error ||= credentials.each_value.collect do |val|
+      "credential value must be string or a hash" unless val.kind_of?(String) || val.kind_of?(Hash)
+      "credential value must have credential_ref and credential_field" if val.kind_of?(Hash) && (!val.key?("credential_ref") || !val.key?("credential_field"))
```

Comment:
the first error will not go through

Think you need a block

```suggestion
      if val.kind_of?(Hash)
        unless val.key?("credential_ref") && val.key?("credential_field")
          "credential value must have credential_ref and credential_field"
        end
      elsif !val.kind_of?(String)
        "credential value must be string or a hash"
      end
```